packages/components/src/components/Container/Container.tsx

```diff
@@ -1,25 +1,36 @@
-import React, { ReactNode } from 'react';
-import { getElementType } from '../helpers';
+import React from 'react';
 import cx from 'clsx';
 
-type ContainerProps = {
+/**
+ * Props Types for the Container component.
+ * They are able to infer the props of the element type passed to the `as` prop.
+ */
+type ContainerProps<T extends React.ElementType> = {
   /** Primary content. */
-  children: ReactNode;
+  children: React.ReactNode;
   /** An element type to render as (string or function). */
-  as: ReactNode;
+  as?: T;
   /** Additional classes. */
-  className: string;
+  className?: string;
   /** Layout size */
-  layout: boolean;
+  layout?: boolean;
   /** Narrow size. */
-  narrow: boolean;
-};
+  narrow?: boolean;
+} & React.ComponentPropsWithoutRef<React.ElementType extends T ? 'div' : T>;
 
-export const Container = (props: ContainerProps) => {
-  const { children, className, layout, narrow, ...rest } = props;
-  const classes = cx('a', 'container', className, { layout, narrow });
+export const Container = <T extends React.ElementType = 'div'>(
+  props: ContainerProps<T>,
+) => {
+  const {
+    as: Component = 'div',
+    children,
+    className,
+    layout,
+    narrow,
+    ...rest
+  } = props;
+  const classes = cx('q', 'container', className, { layout, narrow });
```

Comment:
q for quanta?

I would spell it out, tbh. Too much risk of colliding with some random library that decided to use single-letter class names.